### PR TITLE
:cop: Throttle generation of auth tokens

### PR DESCRIPTION
**Pull request by jaikoo**
Chances of collisions via token entropy attacks are extremely small just because of how they're generated. Still, in order to discourage this kind of attack and any DDOS by side effect, a limit is now set on how many tokens can be generated within an hourly period per person.

The default limit is 8 and can be increased/decreased by setting the env variable MAX_TOKENS_PER_HOUR on the server.

### CODE REVIEW

File: app/models/token.rb [hunk 1]
```diff
@@ -52,4 +68,8 @@ def deactivate_tokens
   def generate_value
     self.value ||= SecureRandom.uuid
   end
+
+  def tokens_in_the_last_hour
+    Token.where(['user_email = ? and created_at between ? and ?', user_email, 1.hour.ago, Time.now])
```
Comment (threedaymonk) [inline app/models/token.rb]: I think it might be nicer to use Arel expressions rather than raw SQL.

File: app/controllers/tokens_controller.rb [hunk 2]
```diff
@@ -58,4 +60,9 @@ def ttl_seconds_in_hours
     hours = minutes.div(60)
     hours
   end
+
+  def throttle_limit_error
```
Comment (threedaymonk) [inline app/controllers/tokens_controller.rb]: I found this name a bit hard to parse. Might something like `show_throttle_limit_error` be clearer?

File: app/controllers/tokens_controller.rb [hunk 3]
```diff
@@ -3,6 +3,8 @@ class TokensController < ApplicationController
   before_action :set_desired_path, only: [:show]
   before_action :ensure_token_auth_enabled!
 
+  rescue_from Token::MaxNumberOfTokensError, with: :throttle_limit_error
```
Comment (threedaymonk) [inline app/controllers/tokens_controller.rb]: I'm getting flashbacks to `On Error GoTo`…

Comment (jaikoo) [inline app/controllers/tokens_controller.rb]: Yeh, I know what you mean but I wanted to have something declarative around catching that particular error rather than wrapping the code in a rescue block.

File: app/models/token.rb (hunk 1 above)
Comment (jaikoo) [inline app/models/token.rb]: Sure, ok. I guess I'm just showing my age ;).

File: app/controllers/tokens_controller.rb (hunk 2 above)
Comment (jaikoo) [inline app/controllers/tokens_controller.rb]: Yup agreed.

**Comment by jaikoo** [comment]
@threedaymonk can you take a look see at my refactoring providings? :)

**Comment by jaikoo** [comment]
Yup. you're right. But if you look at the original code that was there before, it looks like the case of invalid email was only handled in the fact that it re-render the 'new' action and didn't output any error messaging around it leaving the user confused.

I'll setup a separate story in trello to handle error messages for invalid emails addresses. It's a small change.

**Comment by jaikoo** [comment]
Actually I might be wrong... The there's a feature that says it should output an error message but for some reason it's not failing. Will investigate.

**Comment by jaikoo** [comment]
@threedaymonk yeh you're right. I'd previously read the code wrong. Can you take another look and the fix?

File: spec/models/tokens_spec.rb [hunk 4]
```diff
@@ -93,4 +93,20 @@
...
+    let(:person) { create(:person, email: 'text.user@digital.justice.gov.uk') }
+
+    describe '#save or #create of a new token' do
+      it 'throws and error if more than 8 tokens have been generated in the past hour for the same person' do
+        1.upto(9) do |count|
+          if count < 9
+            expect { described_class.for_person(person) }.to_not raise_error
+          else
+            expect { described_class.for_person(person) }.to raise_error
+          end
```
Comment (threedaymonk) [inline spec/models/tokens_spec.rb]: I think this is fine, but I'd have written it like this:

```ruby
8.times do
  expect { described_class.for_person(person) }.to_not raise_error
end
expect { described_class.for_person(person) }.to raise_error
```

File: app/controllers/tokens_controller.rb [hunk 5]
```diff
@@ -4,10 +4,11 @@ class TokensController < ApplicationController
   before_action :ensure_token_auth_enabled!
 
   def create
-    @token = Token.new(token_params)
-    if @token.save
-      TokenMailer.new_token_email(@token).deliver_later
-      render
+    @token = generate_token(token_params)
+    if @token.nil?
+      show_throttle_limit_error
```
Comment (threedaymonk) [inline app/controllers/tokens_controller.rb]: This still shows the throttle limit error if the email is invalid (i.e. if `Token.create` fails), doesn't it? I think using `nil` to mean throttled is inherently problematic - at least if we want to distinguish error cases on the front end.

File: app/controllers/tokens_controller.rb [hunk 6]
```diff
@@ -4,10 +4,9 @@ class TokensController < ApplicationController
   before_action :ensure_token_auth_enabled!
 
   def create
-    @token = Token.new(token_params)
+    @token = Token.create(token_params)
     if @token.save
```
Comment (threedaymonk) [inline app/controllers/tokens_controller.rb]: This will save the token twice now that the previous line is `create`, no?

**Comment by threedaymonk** [comment]
@jaikoo Looks good. Can you squash and rebase onto master?

**Comment by jaikoo** [comment]
Yup, I'll do it now. Thanks!
